---
title: "CTL_ActD_proteomics_analysis"
author: "Arianne Richard"
date: "4 Feb 2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

This script will analyse data-independent acquisigion (DIA) mass spec proteomics data generated by Julia Marchingo (Doreen Cantrell lab, Dundee) in Feb-May 2020 from samples submitted by Arianne Richard (Gillian Griffiths lab, Cambridge). Samples are day-7 cytotoxic T lymphocytes (CTLs) cultured in 6-well plates that were coated or not with anti-CD3 antibodies, with the addition of actinomycin D (ActD) or DMSO vehicle control, for 4 hours. There are 4 replicates corresponding to cells from 4 separate mice, each of which was stimulated in all conditions.

We would like to understand the effects of ActD treatment on the proteome of the cells. Analyses should look at the impact of this drug in unstimulated and anti-CD3-stimulated cells, taking into account mouse of origin.

Processing of the raw data was performed by collaborators in Dundee. Peptides were mapped using Spectronaut by searching against a library generated from multiple CD8 T cell populations and stimulation conditions. These un-normalized results are designated as "raw.PG.Quantity" in the output file. 

# Normalisation

## Distribution examination

First we read in the data as downloaded from the PRIDE repository (see README) and narrow to the relevant samples and data formats.

```{r readin}
library(readxl)
sample_map <-data.frame(read_excel('experimental_design.xlsx'))

sample_map$name <- sub('JM-DIA-', '', sample_map$Sample.name)
sample_map$name <- sub('.raw', '', sample_map$name)

sample_map$cond <- sample_map$condition
sample_map$cond[sample_map$cond == 'aCD3 + DMSO'] <- '3DMSO'
sample_map$cond[sample_map$cond == 'aCD3 + ActD'] <- '3ActD'
sample_map$cond <- factor(sample_map$cond, levels=c('DMSO', 'ActD', '3DMSO', '3ActD'))

sample_map$replicate <- factor(sample_map$bio.rep)

## samples 1 and 2 are male, while 3 and 4 are female, so let's add that information
sex <- as.character(sample_map$replicate)
sex[sex %in% c('1', '2')] <- 'M'
sex[sex %in% c('3', '4')] <- 'F'
sample_map$sex <- sex
rownames(sample_map) <- sample_map$name

full <- read.csv('JM-DIA_v14_report.csv', header=TRUE)

raw <- full[,c(38:53)]
colnames(raw) <- gsub('.raw.PG.Quantity', '', colnames(raw))
colnames(raw) <- gsub('X.[^.]+..JM.DIA.', '', colnames(raw))
colnames(raw) <- gsub('Sample.', 'Sample-', colnames(raw))

# convert to numeric dataframe
raw <- as.data.frame(sapply(raw, as.numeric))

# missing values are NA in this data
# these are proteins that were detected in at least one sample but not in the given sample. To all intents and purposes, these are 0 and so we will replace them with 0 for now and carry out a more sophisticated lower-limit-of-detection replacement before differential abundance testing later.
for(i in 1:ncol(raw)){
raw[,i][which(is.na(raw[,i]))] <- 0
}
raw <- raw[,sample_map$name]
rownames(raw) <- paste0('X', as.character(c(1:nrow(raw))))

qc <- full[,c(22:37)]
colnames(qc) <- gsub('.raw.PG.NrOfStrippedSequencesUsedForQuantification', '', colnames(qc))
colnames(qc) <- gsub('X.[^.]+..JM.DIA.', '', colnames(qc))
colnames(qc) <- gsub('Sample.', 'Sample-', colnames(qc))
qc <- qc[,sample_map$name]
rownames(qc) <- rownames(raw)

prot_info <- full[,c(1:5)]
rownames(prot_info) <- rownames(raw)

raw <- raw[,rownames(sample_map)]
qc <- qc[,rownames(sample_map)]


library(SummarizedExperiment)
pset <- SummarizedExperiment(assays=list(raw=raw, qc=qc), colData=sample_map, rowData=prot_info)

```

Now we want to look at proteins detected per cell to see if there are any overt biases from ActD treatment or anti-CD3 stimulation. Of note, we use a pseudocount of 1 when we take the log2 transformation.


```{r raw_data}
## reorder pset columns for ease
pset <- pset[,order(pset$replicate)]
pset <- pset[,order(pset$cond)]
colnames(pset)

boxplot(log2(assays(pset)$raw+1),las=3, ylab='raw intensity', col=rep(c('blue', 'purple', 'red', 'orange'), each=4))

```

These look pretty similar and no worrying trends emerge. 


## Normalisation for raw values

In order to perform differential abuandance analysis, we will median-centre the log-transformed raw intensity data. (Note that we add a pseudocount of 1 for log-transformation.)

```{r normalise}

raw_mc <- log2(assays(pset)$raw+1)
meds <- apply(raw_mc, 2, median)
raw_mc <- apply(raw_mc, 2, function(x){
  x-median(x)+mean(meds)
})

assays(pset, withDimnames=FALSE)$raw_medcent <- data.frame(raw_mc) ## had to override dimnames because altered in making data.frame to '.' instead of '-'

boxplot(assays(pset)$raw_medcent,las=3, ylab='median-centered raw intensity', col=rep(c('blue', 'purple', 'red', 'orange'), each=4))


```


# Differential abundance testing

## Linear model 

Of note, 0s pose a particular issue in MS data due to both random and non-random occurences, and the difficulty of separating those. However, for DIA as we have here, it seems more likely to be non-random 0s, or "left censoring". Detection limits can be extremely protein-specific, so there is no easy way to understand how a zero in one protein compares to that in another.

To address this, we'll run a conservative analysis where any 0 is replaced based on the lowest value detected in the dataset. We'll use choose these values from a normal distribution with mean 5% below the lowest detected value and standard deviation the same 5%. This is extremely conservative, but will allow us to run differential abundance tests across the board and will pick up proteins where e.g. condition A has high abundance and condition B has low or 0 abundance. 

Finally, we'll separately detect proteins that are all 0s in one condition and all detected in the other as these (may) represent complete on/off scenarios.

```{r replace_0s}
raw_no_0 <- assays(pset)[['raw']]
sums <- apply(raw_no_0, 1, sum)
raw_no_0 <- raw_no_0[-which(sums==0),] ## 3 rows removed

raw_no_0_rep <- apply(raw_no_0, 1, function(x){
  non0 <- x[!x==0]
  mi <- min(non0)
  mi5 <- mi-0.05*mi
  zeros <- which(x ==0)
  if(length(zeros) > 0){
    set.seed(100)
    x[zeros] <- rnorm(length(zeros), mean=mi5, sd=0.05*mi)
  }
  return(x)
})

raw_no_0_rep <- data.frame(t(raw_no_0_rep))
boxplot(log2(raw_no_0_rep+1),las=3, ylab='log2 raw intensity', col=rep(c('blue', 'purple', 'red', 'orange'), each=4))

## use old median factors for normalization (per sample, so removal of 3 rows is not a problem)

raw_no_0_rep_mc <- log2(raw_no_0_rep +1)
for(i in 1:ncol(raw_no_0_rep)){
  raw_no_0_rep_mc[,i] <- raw_no_0_rep_mc[,i]-meds[i]+mean(meds)
}

boxplot(raw_no_0_rep_mc,las=3, ylab='log2 no 0 median center intensity', col=rep(c('blue', 'purple', 'red', 'orange'), each=4))

pset_no0 <- SummarizedExperiment(assays=list(raw=raw_no_0_rep, raw_medcent=raw_no_0_rep_mc, qc=assays(pset)$qc[-which(sums==0),]), colData=colData(pset), rowData=prot_info[-which(sums==0),])

```

Now let's make a QC flag column to mark proteins that are detected by single peptides (or no peptides) in more than 75% of samples as these are more likely to be false. 

```{r flags}

Singleflag <- assays(pset_no0)$qc
Singleflag <- apply(Singleflag, 1, function(x){
  length(which(x %in% c('1', 'Filtered'))) <= length(x)*0.75
})
table(Singleflag)
rowData(pset_no0)$Singleflag <- Singleflag

```

Now we can test for differential abundance. We will remove our single-peptide hits.

```{r DA_normalized_remove_singles}

DA_test <- function(pset, assay_tab, condition1, condition2, filt=rep(TRUE, nrow(pset)), flag=rep(TRUE, nrow(pset))){
  ds <- pset
  ds <- ds[filt,]
  ds$cond <- factor(ds$cond, 
                    levels=c(condition1, condition2, levels(ds$cond)[!levels(ds$cond) %in% c(condition1, condition2)]))
  stats <- apply(assays(ds)[[assay_tab]], 1, function(x){
    lm(x ~ ds$replicate + ds$cond)
  })
  est <- sapply(stats, function(x){summary(x)$coefficients[5,1]})
  se <- sapply(stats, function(x){summary(x)$coefficients[5,2]})
  ps <- sapply(stats, function(x){summary(x)$coefficients[5,4]})
  BH <- p.adjust(ps, method='fdr')
  flag <- flag[filt]
  res <- data.frame(prot=rowData(ds)$PG.ProteinAccessions, 
                    gene=rowData(ds)$PG.Genes,
                    desc=rowData(ds)$PG.ProteinDescriptions,
                    estimate=est, std.err=se, 
                    conf.int.95.bot=est-1.96*se, conf.int.95.top=est+1.96*se, 
                    p.value=ps, fdr=BH)
  if(length(flag==FALSE) > 0){
    res$flag <- flag
  }
  return(res)
}

CD3_res_noS <- DA_test(pset_no0, 'raw_medcent', condition1='3DMSO', condition2='3ActD', filt=rowData(pset_no0)$Singleflag)

write.table(CD3_res_noS, 'plots_and_tables/stimulated_ActDvDMSO_linearmodel_replace0s_removeSingles.txt', sep='\t', quote=FALSE)

null_res_noS <- DA_test(pset_no0, 'raw_medcent', condition1='DMSO', condition2='ActD', filt=rowData(pset_no0)$Singleflag)

write.table(null_res_noS, 'plots_and_tables/unstimulated_ActDvDMSO_linearmodel_replace0s_removeSingles.txt', sep='\t', quote=FALSE)

stim_v_unstim_res_noS <- DA_test(pset_no0, 'raw_medcent', condition1='DMSO', condition2='3DMSO', filt=rowData(pset_no0)$Singleflag)

write.table(stim_v_unstim_res_noS, 'plots_and_tables/stimulatedvunstimulated_linearmodel_replace0s_removeSingles.txt', sep='\t', quote=FALSE)

```

And we'll make some volcano plots of the results.

```{r volcano_plots}

cols <- rep('black', times=nrow(CD3_res_noS))
cols[CD3_res_noS$fdr < 0.05] <- 'darkred'
cols[(CD3_res_noS$fdr < 0.05) & (abs(CD3_res_noS$estimate) > 1)] <- 'red'

png('plots_and_tables/volcano_stimAvD_replace0s_noSingles.png', height=3, width=3.3, unit='in', res=300)
par(cex=0.5, bty='l', mgp=c(1,1,0), mar=c(3,3,0,1), cex.axis=2)
plot(CD3_res_noS$estimate, -log10(CD3_res_noS$p.value), pch=16, col=cols, 
     xlab='', ylab='', 
     xlim=c(-17, 17), ylim=c(0, 9),
     )
dev.off()


cols <- rep('black', times=nrow(null_res_noS))
cols[null_res_noS$fdr < 0.05] <- 'darkred'
cols[(null_res_noS$fdr < 0.05) & (abs(null_res_noS$estimate) > 1)] <- 'red'

png('plots_and_tables/volcano_unstimAvD_replace0s_noSingles.png', height=3, width=3.3, unit='in', res=300)
par(cex=0.5, bty='l', mgp=c(1,1,0), mar=c(3,3,0,1), cex.axis=2)
plot(null_res_noS$estimate, -log10(null_res_noS$p.value), pch=16, col=cols, 
     xlab='', ylab='', 
     xlim=c(-17, 17), ylim=c(0, 9),
     )
dev.off()



```

## Comparing effect sizes

Are proteins that increase in abundance with stimulation particularly impacted by actinomycin D treatment? Let's check by comparing effect sizes.

```{r effect_size_comp}

## we want to label some points - this script has been modified for optimal label positioning after initial run
rns <- which(stim_v_unstim_res_noS$estimate > 4)
poses <- rep(1, times=length(rns))
poses[which(stim_v_unstim_res_noS$gene[rns] =='Ifng')] <- 3
poses[which(stim_v_unstim_res_noS$gene[rns] =='Csf2')] <- 4
poses[which(stim_v_unstim_res_noS$gene[rns] =='Itga3')] <- 4
poses[which(stim_v_unstim_res_noS$gene[rns] =='Xpa')] <- 3
poses[which(stim_v_unstim_res_noS$gene[rns] =='Egr1')] <- 2


textcol <- rep('black', length(rns))

cols <- rep('grey60', times=nrow(stim_v_unstim_res_noS))
cols[which(stim_v_unstim_res_noS$fdr < 0.05)] <- 'deepskyblue'
cols[which(CD3_res_noS$fdr < 0.05)] <- 'springgreen3'
cols[which((stim_v_unstim_res_noS$fdr < 0.05) & (CD3_res_noS$fdr < 0.05))] <- 'deeppink'

png('plots_and_tables/effectsizes_stimvunstim_CD3AvD_replace0s_noSingles.png', height=4, width=4, unit='in', res=300)
par(cex=0.5, bty='l', mgp=c(3, 1, 0), mar=c(6, 6, 2, 2), cex.axis=1.8, cex.lab=2.2)
plot(stim_v_unstim_res_noS$estimate, CD3_res_noS$estimate, pch=16, col=cols, xlab='Protein: stimulated v resting', ylab='Protein: stimulated ActD v DMSO')
text(stim_v_unstim_res_noS$estimate[rns], CD3_res_noS$estimate[rns], labels=toupper(CD3_res_noS$gene[rns]), pos=poses, col=textcol, cex=1.2)
abline(h=0, lty=3)
abline(v=0, lty=3)
dev.off()

```

And what does this look like if we consider how the transcriptome changes upon restimulation? Let's compare the effect sizes of transcriptional changes with stimulation (previous RNAseq experiment using 60m v unstimulated) with protein changes at 4h upon transcription blockade. 

```{r RNAseq_effect_size_comparison}

a <- read.table('../RNAseq/plots_and_tables/DE_60m.txt')

a <- a[a$symbol %in% CD3_res_noS$gene,]
nu <- a$symbol[which(duplicated(a$symbol))]
a <- a[!a$symbol %in% nu,]
b <- CD3_res_noS[CD3_res_noS$gene %in% a$symbol,]
nu <- b$gene[which(duplicated(b$gene))]
b <- b[!b$gene %in% nu,]
df <- merge(a, b, by.x='symbol', by.y='gene')

## and we want to label some points - again, positions have been optimized after initial run
rns <- which(df$logFC > 3)
poses <- rep(1, times=length(rns))
poses[which(df$symbol[rns] =='Ifng')] <- 3
poses[which(df$symbol[rns] =='Tnfsf9')] <- 3
poses[which(df$symbol[rns] =='Tnf')] <- 3
poses[which(df$symbol[rns] =='Nr4a1')] <- 4
poses[which(df$symbol[rns] =='Nr4a2')] <- 4
poses[which(df$symbol[rns] =='Csf2')] <- 4
poses[which(df$symbol[rns] =='Egr3')] <- 4
textcol <- rep('black', length(rns))

cols <- rep('grey60', times=nrow(df))
cols[which(df$FDR < 0.05)] <- 'deepskyblue'
cols[which(df$fdr < 0.05)] <- 'springgreen3'
cols[which((df$FDR < 0.05) & (df$fdr < 0.05)) ] <- 'deeppink'

make.italic <- function(x) {as.expression(lapply(x, function(y) bquote(italic(.(y)))))}

png('plots_and_tables/effectsizes_RNAstimvunstim_CD3AvD_replace0s_noSingles.png', height=4, width=4, unit='in', res=300)
par(cex=0.5, bty='l', mgp=c(3, 1, 0), mar=c(6, 6, 2, 2), cex.axis=1.8, cex.lab=2.2)
plot(df$logFC, df$estimate, pch=16, col=cols, xlab='RNA: stimulated v resting', ylab='Protein: stimulated ActD v DMSO')
text(df$logFC[rns], df$estimate[rns], labels=make.italic(df$symbol[rns]), pos=poses, col=textcol, cex=1.2)
abline(h=0, lty=3)
abline(v=0, lty=3)
dev.off()

```

And how does ActD treatment compare between stimulated and unstimulated cells? Start with a correlation test.

```{r ActD_effect_size_comp}

cor.test(CD3_res_noS$estimate, null_res_noS$estimate, method='spearman')

```

Now, let's look at the genes that are downregulated by actinomycin D in both resting and stimulated CTLs and those genes that are unique to stimulated CTLs. We'll define the "both" category by FDR < 0.05 and logFC < -1 in both. We'll define the "unique to stimulated" category by FDR < 0.05 and logFC < -1 in stimulated cells and 95% confidence interval crossing 0 in resting cells.

```{r}

both <- which((null_res_noS$fdr < 0.05) & (null_res_noS$estimate < -1) & 
                (CD3_res_noS$fdr < 0.05) & (CD3_res_noS$estimate < -1))
both_mat <- assays(pset_no0)[['raw_medcent']][rowData(pset_no0)$Singleflag,][both,]
rownames(both_mat) <- toupper(rowData(pset_no0)$PG.Genes[rowData(pset_no0)$Singleflag][both])
## need to reorder to make figures in manuscript comparable
ord <- as.numeric(sub('Sample-', '', colnames(both_mat)))
both_mat <- both_mat[,ord]

unique <- which((null_res_noS$conf.int.95.bot < 0) & (null_res_noS$conf.int.95.top > 0) & 
                (CD3_res_noS$fdr < 0.05) & (CD3_res_noS$estimate < -1))

unique_mat <- assays(pset_no0)[['raw_medcent']][rowData(pset_no0)$Singleflag,][unique,]
rownames(unique_mat) <- toupper(rowData(pset_no0)$PG.Genes[rowData(pset_no0)$Singleflag][unique])
unique_mat <- unique_mat[,ord]
## fix missing gene name 
null_res_noS[unique[which(rownames(unique_mat)=='')],]
rownames(unique_mat)[which(rownames(unique_mat)=='')] <- 'TCRB1/2 chain C region'

col_annot <- as.character(pset_no0$cond)
col_annot[col_annot == '3DMSO'] <- 'antiCD3_DMSO'
col_annot[col_annot == '3ActD'] <- 'antiCD3_ActD'
col_annot <- col_annot[ord]
col_annot <- data.frame(col_annot)
colnames(col_annot) <- c('Condition')
rownames(col_annot) <- colnames(both_mat)

## determine max and min for legend

maxleg <- max(c(max(abs(scale(t(both_mat)))), max(abs(scale(t(unique_mat))))))

library(pheatmap)

pheatmap(as.matrix(both_mat), cluster_cols=FALSE, show_colnames=FALSE, scale='row', 
         breaks=seq(-maxleg, maxleg,length.out=100), cellwidth=8, cellheight=9,
         annotation_col=col_annot, annotation_names_col=FALSE, annotation_legend=FALSE, legend=FALSE,
         annotation_colors=list(Condition=c(DMSO='blue', antiCD3_DMSO='red', ActD='purple', antiCD3_ActD='orange')),
         filename='plots_and_tables/ActDvDMSO_both_HM.png')



pheatmap(as.matrix(unique_mat), cluster_cols=FALSE, show_colnames=FALSE, scale='row', 
         breaks=seq(-maxleg, maxleg,length.out=100), cellwidth=8, cellheight=9,
         annotation_col=col_annot, annotation_names_col=FALSE, 
         annotation_colors=list(Condition=c(DMSO='blue', antiCD3_DMSO='red', ActD='purple', antiCD3_ActD='orange')),
         filename='plots_and_tables/ActDvDMSO_uniqueCD3_HM.png')

png('plots_and_tables/effectsizes_CD3AvD_nullAvD_replace0s_noSingles.png', height=3, width=3, unit='in', res=300)
par(cex=0.5, bty='l', mgp=c(3, 1, 0), mar=c(6, 6, 2, 2), cex.axis=1.8, cex.lab=2.2)
plot(null_res_noS$estimate, CD3_res_noS$estimate, col='black', pch=16, xlab='resting ActD v DMSO', ylab='stimulated ActD v DMSO')
points(null_res_noS$estimate[both], CD3_res_noS$estimate[both],col='magenta', pch=16)
points(null_res_noS$estimate[unique], CD3_res_noS$estimate[unique],col='turquoise', pch=16)
abline(h=0, lty=3)
abline(v=0, lty=3)
dev.off()

```

It looks like cytokine upregulation is particularly affected by actinomycin D treatment. We'd like to compare the effects on cytokines/receptors. For this we need a list of proteins for each. We annotate cytokines and their receptors from KEGG annotation of "Cytokine-cytokine receptor interaction" performed by Cantrell lab in Dundee. These were imported in the original analysis but are manually added in this script so it can be reproduced by other groups.

```{r cytokines_receptors_cytolytic}
#ID'd by KEGG
cytokines_receptors <- which(rowData(pset_no0[rowData(pset_no0)$Singleflag,])$PG.Genes %in% 
                               c("Tnfsf11","Tnfrsf18","Ifng","Il1a","Csf2","Il2ra","Tgfb1","Il2",
                                 "Kit","Tnf","Lif","Lta","Ccl1","Ccl3","Ccl4","Ifngr1","Il2rb","Il4r",
                                 "Il7r","Il10","Tnfrsf9","Tnfrsf1b","Il3ra","Il2rg","Cd27","Tnfsf9",
                                 "Tnfrsf4","Ccr7","Xcl1","Cxcr4","Il12rb1","Tnfrsf8","Il18r1","Tgfbr2",
                                 "Crlf2","Il24","Tnfsf14"))
rowData(pset_no0[rowData(pset_no0)$Singleflag,])$PG.Genes[cytokines_receptors]
## manual filtering cytokines and receptors from output
cytokines <- cytokines_receptors[c(1,3,4,5,7,8,10,11,12,13,14,15,20,26,29,36,37)]
receptors <- cytokines_receptors[!cytokines_receptors %in% cytokines]

rowData(pset_no0[rowData(pset_no0)$Singleflag,])$PG.Genes[cytokines]
rowData(pset_no0[rowData(pset_no0)$Singleflag,])$PG.Genes[receptors]

## and we define cytolytic genes as granzymes or perforin
cytolytic <- c(grep('Gzm', rowData(pset_no0[rowData(pset_no0)$Singleflag,])$PG.Genes), 
               grep('Prf1', rowData(pset_no0[rowData(pset_no0)$Singleflag,])$PG.Genes)) 
rowData(pset_no0[rowData(pset_no0)$Singleflag,])$PG.Genes[cytolytic]

effects <- list(cytokines=CD3_res_noS$estimate[cytokines], receptors=CD3_res_noS$estimate[receptors], cytolytic=CD3_res_noS$estimate[cytolytic])
top <- ceiling(max(sapply(effects, max)))
bot <- floor(min(sapply(effects, min)))

pdf('plots_and_tables/cytokines_cytolytic.pdf', height=3.5, width=4)
par(bty='l')
boxplot(effects, outline=FALSE, ylim=c(bot, top), ylab=expression('log'[2]*' (fold-change)'))
stripchart(effects, add=TRUE, method='jitter', pch=1, col='black', vertical=TRUE, cex=1)
dev.off()

```

And finally, what does stimulation do to the proteome when cells cannot transcribe?

```{r DA_normalized_ActD}
ActD_stim_v_unstim_res_noS <- DA_test(pset_no0, 'raw_medcent', condition1='ActD', condition2='3ActD', filt=rowData(pset_no0)$Singleflag)

write.table(ActD_stim_v_unstim_res_noS, 'plots_and_tables/ActDstimulatedvunstimulated_linearmodel_replace0s_removeSingles.txt', sep='\t', quote=FALSE)

```

Very very little, and mostly down. Interesting. Transcription is required for basically all TCR-induced expression at 4h.

## Enrichment testing

Let's look for enrichment of protein classes among proteins that are substantially downregulated under anti-CD3 stimulation with actinomycin D treatment compared to DMSO. 

We filtered for proteins with FDR < 0.05 and log-FC < -1 externally and then ran them through PANTHER Overrepresentation Test comparing to a background of all proteins included in the differential abundance test using a Fisher's Exact Test and FDR < 0.05. Analysis was run on 9/8/2022. We exported the results and now read them in below to plot.

```{r enrichment}

enrich <- read.table('plots_and_tables/stimulated_ActDvDMSO_FDR0.05_2xDOWN_panther_enrichment.txt', sep='\t', header=TRUE, skip=11)

enrich <- enrich[order(enrich$Client.Text.Box.Input..fold.Enrichment.),]
enrich$namessimp <- unlist(lapply(strsplit(as.character(enrich$PANTHER.Protein.Class), split='(', fixed=TRUE), function(x) {x[1]}))

pdf('plots_and_tables/stimulated_ActDvDMSO_FDR0.1_2xDOWN_panther_enrichment.pdf', height=3, width=6)
par(mar=c(5, 15, 2, 3)+0.1, cex.axis=0.8, cex.lab=1, mgp=c(1.5, 0.5, 0), font.main=1)
barplot(height=enrich$Client.Text.Box.Input..fold.Enrichment., names.arg=as.character(enrich$namessimp), 
        horiz=TRUE, xlab='fold enrichment', col='grey', las=1, 
        main='')
dev.off()

```


## Testing on/off proteins

And we will have a final look at the proteins with 0 counts. We're looking for proteins that are "on" in one condition and "off" in the other. 

We'll once again remove those proteins detected by single peptides. Because we want to look at proteins that are completely absent in one condition, we'll make a strict requirement that the protein is always detected by more than 1 peptide in the other condition. 

```{r on_off}

onoff_test <- function(pset, assay_tab, condition1, condition2){
  ds <- pset[,pset$cond %in% c(condition1, condition2)]
  ds$cond <- factor(as.character(ds$cond), levels=c(condition1, condition2))
  ds <- ds[,order(ds$cond, ds$replicate)]
  onoff <- apply(assays(ds)[['raw']], 1, function(x){
    if(length(which(x[which(ds$cond == condition1)] > 0)) == length(x[which(ds$cond == condition1)])){
      if(length(which(x[which(ds$cond == condition2)] == 0)) == length(x[which(ds$cond == condition2)])){
        return('on')
      }else{
        return('-')
      }
    }else if(length(which(x[which(ds$cond == condition1)] == 0)) == length(x[which(ds$cond == condition1)])){
      if(length(which(x[which(ds$cond == condition2)] > 0)) == length(x[which(ds$cond == condition2)])){
        return('off')
      }else{
        return('-')
      }
    }else{
      return('-')
    }
  })
  onoff <- unlist(onoff)

  singles <- assays(ds)$qc
  singles <- apply(singles, 1, function(x){
    if(length(which(x[which(ds$cond == condition1)] == 1)) == length(x[which(ds$cond == condition1)])){
      return('single')
    }else if(length(which(x[which(ds$cond == condition2)] == 1)) == length(x[which(ds$cond == condition2)])){
      return('single')
    }else{
      return('-')
    }
  })

  res <- data.frame(prot=rowData(ds)$PG.ProteinAccessions, 
                    gene=rowData(ds)$PG.Genes,
                    desc=rowData(ds)$PG.ProteinDescriptions,
                    onoff=onoff, 
                    singles=singles)
  res <- res[!res$singles == 'single',]
  return(res)
}

CD3oo <- onoff_test(pset, assay_tab='raw_medcent', condition1='3ActD', condition2='3DMSO')
CD3oo[CD3oo$onoff %in% c('on', 'off'),]

write.table(CD3oo, 'plots_and_tables/CD3_ActDvDMSO_noS_onoff.txt', sep='\t', quote=FALSE)

nulloo <- onoff_test(pset, assay_tab='raw_medcent', condition1='ActD', condition2='DMSO')
nulloo[nulloo$onoff %in% c('on', 'off'),]

write.table(nulloo, 'plots_and_tables/null_ActDvDMSO_noS_onoff.txt', sep='\t', quote=FALSE)

```

Only Ier5 and Egr1 came up in stimulated cells, and Ier5 alone in unstimulated. The stimulated cell proteins were already captured as differentially abundant by our 0-replacement method, while Ier5 in unstimulated cells was found but not significant at FDR < 0.05. These are very few, and our conclusions are really only about the affect of ActD on stimulated cells, so this analysis is basically not necessary. We haven't missed anything important through our 0-replacement method.

# Plots

Finally, we'd also like to make some plots of individual proteins for figures.

```{r plots}

plot_prot <- function(pset, geneID, assay='raw_medcent', prelog=TRUE, 
                      conditions=c('DMSO', '3DMSO', 'ActD', '3ActD'), 
                      condition_labels=c('DMSO', 'anti-CD3 + DMSO', 'ActD', 'anti-CD3 + ActD'), 
                      cols=c('white')){
  require(ggplot2)
  rn <- rownames(pset)[which(rowData(pset)$PG.Genes == geneID)]
  print(rn)
  dat1 <- as.numeric(assays(pset)[[assay]][rn,])
  names(dat1) <- colnames(pset)
  dat1 <- split(dat1, pset$cond)
  dat1 <- dat1[names(dat1) %in% conditions]
  dat1 <- dat1[conditions]
  dat2 <- vector('list', length(dat1))
  for(i in 1:length(dat1)){
    dat2[[i]] <- data.frame(cond=rep(names(dat1)[i], length(dat1[[i]])), expr=dat1[[i]])
  }
  dat2 <- do.call('rbind', dat2)
  dat2$cond <- factor(dat2$cond, levels=conditions)
  dat2$mouse <- factor(as.numeric(colData(pset)[rownames(dat2), 'replicate']))
  if(prelog==FALSE){
    dat2$expr <- log2(dat2$expr +1)
  }
  p <- ggplot(dat2, aes(x=cond, y=expr, fill=mouse)) +
    geom_boxplot(fill=cols, outlier.shape = NA) +
    geom_dotplot(binaxis='y', stackdir='center', position=position_dodge(0.2)) +
    scale_fill_grey() +
    scale_x_discrete(breaks=conditions,
        labels=condition_labels) + 
    theme_classic() +
    theme(
    axis.text.x = element_text(color="black"),
    axis.text.y = element_text(color="black"),
    axis.ticks = element_line(color = "black")
  )

  return(p)
}

pdf('plots_and_tables/Xcl1.pdf', height=3.5, width=3)
p <- plot_prot(pset=pset, geneID='Xcl1', assay='raw_medcent', prelog=TRUE)
p + 
  labs(x ="", y = expression('log'[2]*' XCL1 abundance')) + 
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  scale_y_continuous(limits = c(-0.5, 30))
dev.off()

pdf('plots_and_tables/Ifng.pdf', height=3.5, width=3)
p <- plot_prot(pset=pset, geneID='Ifng', assay='raw_medcent', prelog=TRUE)
p + 
  labs(x ="", y = expression('log'[2]*' IFNG abundance')) + 
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  scale_y_continuous(limits = c(-0.5, 30))
dev.off()


pdf('plots_and_tables/Gzmb.pdf', height=3.5, width=3)
p <- plot_prot(pset=pset, geneID='Gzmb', assay='raw_medcent', prelog=TRUE)
p + 
  labs(x ="", y = expression('log'[2]*' GZMB abundance')) + 
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  scale_y_continuous(limits = c(-0.5, 30))
dev.off()

pdf('plots_and_tables/Prf1.pdf', height=3.5, width=3)
p <- plot_prot(pset=pset, geneID='Prf1', assay='raw_medcent', prelog=TRUE)
p + 
  labs(x ="", y = expression('log'[2]*' PRF1 abundance')) + 
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) +
  scale_y_continuous(limits = c(-0.5, 30))
dev.off()

```


# Finishing up

```{r}
sessionInfo()
```
